huntsman/observatory.py

```diff
@@ -512,3 +363,159 @@ def _create_flat_field_observation(self, alt=None, az=None,
+                        alt=alt, az=az, filter_name=filter_names[cam_name])
+
+        # Loop until conditions are met to finish flat-fielding
+        exposure_count = 0
+        while True:
+
+            # Update the exposure count
+            exposure_count += 1
+
+            # Get the FITS headers
+            fits_headers = {cam_name: self.get_standard_headers(
+                observation=observations[cam_name]) for cam_name in camera_names}
+
+            # Use a common start time for all cameras
+            start_time = utils.current_time()
+            for cam_name in camera_names:
+                fits_headers[cam_name]['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            _exptimes = {cam_name: exptimes[cam_name][-1] for cam_name in camera_names}
+            camera_events = self._take_flat_fields(
+                        camera_names, observations, _exptimes, fits_headers)
+
+            # Check the counts for each image
+            is_saturated = False
+            for cam_name, info in camera_events.items():
+                filename = info['filename']
+                self.logger.debug("Checking counts for {}".format(filename))
+
+                # Unpack fits if compressed
+                if not os.path.exists(filename) and \
+                        os.path.exists(filename.replace('.fits', '.fits.fz')):
+                    fits_utils.fpack(filename.replace('.fits', '.fits.fz'), unpack=True)
+
+                # Calculate average counts per pixel
+                data = fits.getdata(filename)
+                mean_counts, _, _ = stats.sigma_clipped_stats(data)
+                mean_counts -= bias
+                self.logger.debug("Counts: {}".format(mean_counts))
+                if (mean_counts < min_counts) or (mean_counts > max_counts):
+                    self.logger.debug("Counts outside min/max range, should be discarded.")
+
+                # Check if camera is saturated
+                if mean_counts >= max_counts:
+                    is_saturated = True
+
+                # Get suggested exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                self.logger.debug("Elapsed time: {}".format(elapsed_time))
+                exptime = int(exptimes[cam_name][-1].value * (target_adu / mean_counts) *
+                              (2.0 ** (elapsed_time / 180.0)) + 0.5)
+                exptime = max(1, exptime)
+                self.logger.debug("Suggested exptime for {}: {}".format(cam_name, exptime))
+                exptimes[cam_name].append(exptime * u.second)
+
+            # Stop flats if any time is greater than max
+            if any([t[-1].value >= max_exptime for t in exptimes.values()]) & \
+                    (not self.past_midnight):
+                self.logger.debug("Exposure times greater than max, stopping flat fields.")
+                break
```

Comment:
This made a hard-to-read line less readable. Let's clean it up:

```suggestion
            exptime_too_long = [t[-1].value >= max_exptime for t in exptimes.values()]
            if any(exptime_too_long) and not self.past_midnight:
                self.logger.debug("Exposure times greater than max, stopping flat fields.")
                break
```